### PR TITLE
[Merged by Bors] - feat: Variant of structure theorem of finite abelian groups

### PR DESCRIPTION
**Pull request by YaelDillies**
where the trivial factors have been dropped.

I have the idea of a general approach to dropping trivial terms from a direct sum. However it is too complicated for me to unsorry it and overkill here, so I am leaving it as a comment.

From LeanAPAP

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/ZMod/Basic.lean [hunk 1]
```diff
@@ -854,6 +854,18 @@ def chineseRemainder {m n : ℕ} (h : m.Coprime n) : ZMod (m * n) ≃+* ZMod m 
     right_inv := inv.2 }
 #align zmod.chinese_remainder ZMod.chineseRemainder
 
+lemma subsingleton_iff {n : ℕ} : Subsingleton (ZMod n) ↔ n = 1 := by
```
Comment (jcommelin) [inline Mathlib/Data/ZMod/Basic.lean:857]:
```suggestion
@[simp]
lemma subsingleton_iff {n : ℕ} : Subsingleton (ZMod n) ↔ n = 1 := by
```

File: Mathlib/GroupTheory/FiniteAbelian.lean [hunk 2]
```diff
@@ -14,14 +14,73 @@ import Mathlib.Data.ZMod.Quotient
...
+* `AddCommGroup.equiv_directSum_zmod_of_finite` : Any finite abelian group is a direct sum of
   some `ZMod (p i ^ e i)` for some prime powers `p i ^ e i`.
 
 -/
 
-
 open scoped DirectSum
 
+/-
+TODO: Here's a more general approach to dropping trivial factors from a direct sums:
```
Comment (jcommelin) [inline Mathlib/GroupTheory/FiniteAbelian.lean:25]: Please indicate briefly what it takes to fill in the sorries. Just hard work, or are there prerequisites missing?

File: Mathlib/GroupTheory/FiniteAbelian.lean [hunk 3]
```diff
@@ -14,14 +14,73 @@ import Mathlib.Data.ZMod.Quotient
...
+    (fun j hj hi ↦ rfl)
+    (fun i hi hj a ↦ rfl)
+    (fun j hj hi a ↦ rfl)
+-/
+
+private def myThingForward {ι : Type} [DecidableEq ι] (p : ι → ℕ) (n : ι → ℕ) :
+    (⨁ i : {i // n i ≠ 0}, ZMod (p i ^ n i)) →+ ⨁ i, ZMod (p i ^ n i) :=
+  DirectSum.toAddMonoid fun i ↦ DirectSum.of (fun i ↦ ZMod (p i ^ n i)) i
+
+private def myThing (ι : Type) [DecidableEq ι] (p : ι → ℕ) (n : ι → ℕ) :
```
Comment (jcommelin) [inline Mathlib/GroupTheory/FiniteAbelian.lean:63]: Hmmz, are these the best names?

File: Mathlib/GroupTheory/FiniteAbelian.lean (hunk 2 above)
Comment (YaelDillies) [reply Mathlib/GroupTheory/FiniteAbelian.lean:25]: Just hard work probably

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**